{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a4742cde",
   "metadata": {},
   "source": [
    "CTDC014 ctDNA Analysis \n",
    "\n",
    "How this patient was selected?\n",
    "\n",
    "Using the command - \n",
    "\n",
    "(echo -e \"PatientID\\tSample Name\\tIsolate\\tTreatment Stage\\tExperiment\"; csvcut -c \"Sample Name\",\"isolate\",\"Experiment\" ~/Downloads/SraRunTable.csv | sed '1d' | awk -F',' '{\n",
    "  if ($1 ~ /^PBMC_/) stage=\"Pre-treatment\";\n",
    "  else if ($1 ~ /^FFPE_/) stage=\"Post-treatment\";\n",
    "  else if ($1 ~ /-[0-9]+$/) stage=\"During treatment\";\n",
    "  else stage=\"Baseline\";\n",
    "  split($1, a, /[^0-9]/);\n",
    "  for (i in a) if (a[i] ~ /^[0-9]+$/) {id=a[i]; break}\n",
    "  print id \"\\t\" $1 \"\\t\" $2 \"\\t\" stage \"\\t\" $3\n",
    "}' | sort -k1,1n | grep '^014')\n",
    "\n",
    "Output \\n\n",
    "\n",
    "PatientID       Sample Name     Isolate Treatment Stage Experiment \\n\n",
    "\n",
    "014     C_MB014-2       blood   During treatment        SRX10351577 \\n\n",
    "\n",
    "014     C_MB014-3       blood   During treatment        SRX10351578 \\n\n",
    "\n",
    "014     FFPE_C_MB014    tissue biopsy   Post-treatment  SRX10351726 \\n\n",
    "\n",
    "014     PBMC_C_MB014    blood   Pre-treatment   SRX10351683 \\n\n",
    "\n",
    "014     PBMC_C_MB014-2  blood   Pre-treatment   SRX10351672 \\n\n",
    "\n",
    "Metadata file: SraRunTable.csv was downloaded from https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA714799&o=acc_s%3Aa by clicking 'Metadata' under 'Download'\n",
    "\n",
    "CTDC014 in Table 4 in the paper - https://www.nature.com/articles/s41598-021-95345-4/tables/4\n",
    "has Gene KRAS missense variant with Pathogenic/Likely_pathogenic \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac586c7c",
   "metadata": {},
   "source": [
    "Download raw sequencing files \n",
    "\n",
    "for srr in SRR21014913 SRR21014914 SRR21015062 SRR21015019 SRR21015008; do prefetch \"$srr\" && fasterq-dump \"$srr\" --split-files --outdir /Volumes/T7/GTMBioinfo/raw; done\n",
    "\n",
    "QC of sequencing files\n",
    "\n",
    "commands of fastqc and multiqc used\n",
    "\n",
    "    fastqc /Volumes/T7/GTMBioinfo/raw/*.fastq -o docs/\n",
    "\n",
    "    multiqc docs/ -o docs/\n",
    "\n",
    "    Save the multiqc report as PDF in docs/ folder from html report\n",
    "\n",
    "multi qc report - https://github.com/justin-mbca/GTMBioinfo/blob/main/docs/MultiQCReport.pdf\n",
    "From the report, the quality scores are good in overall and there is high adapter content for a few samples.\n",
    "    Therefore, trimming is performed before alignment.\n",
    "\n",
    "    Command - \n",
    " \n",
    "    mkdir -p /Volumes/T7/GTMBioinfo/trimmed\n",
    "\n",
    "    for id in SRR21014913 SRR21014914 SRR21015062 SRR21015019 SRR21015008; do\n",
    "        fastp \\\n",
    "            -i /Volumes/T7/GTMBioinfo/raw/${id}_1.fastq \\\n",
    "            -I /Volumes/T7/GTMBioinfo/raw/${id}_2.fastq \\\n",
    "            -o /Volumes/T7/GTMBioinfo/trimmed/${id}_1.trimmed.fastq \\\n",
    "            -O /Volumes/T7/GTMBioinfo/trimmed/${id}_2.trimmed.fastq \\\n",
    "            --detect_adapter_for_pe \\\n",
    "            --html /Volumes/T7/GTMBioinfo/qc/${id}_fastp.html \\\n",
    "            --json /Volumes/T7/GTMBioinfo/qc/${id}_fastp.json\n",
    "        done\n",
    "\n",
    "The duplicate rates are medium high across samples.\n",
    "    Therefore, the duplicates will be removed during alignment and variant calling steps."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "002748fb",
   "metadata": {},
   "source": [
    "Alignment \n",
    "# Create GRCh38 reference genome\n",
    "   1. Gene KRAS is in Chromosome 12 - https://www.ncbi.nlm.nih.gov/gene?term=(KRAS%5BGene%20Name%5D)%20AND%20Homo%20sapiens%5BOrganism%5D\n",
    "      1.There for use chromosome 12 as reference of mapping and variant calling to analyze gene KRAS \n",
    "   2. Download chromosome 12 of reference fasta file - wget -P /Volumes/T7/GTMBioinfo/reference/ ftp://ftp.ensembl.org/pub/release-115/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.chromosome.12.fa.gz\n",
    "   3. Extract zip file -  gunzip /Volumes/T7/GTMBioinfo/reference/Homo_sapiens.GRCh38.dna.chromosome.12.fa.gz\n",
    "   4. Create BWA index file - bwa index /Volumes/T7/GTMBioinfo/reference/Homo_sapiens.GRCh38.dna.chromosome.12.fa\n",
    "   5. Create FASTA index - samtools faidx /Volumes/T7/GTMBioinfo/reference/Homo_sapiens.GRCh38.dna.chromosome.12.fa\n",
    "# Alignment\n",
    "   1. create bash script in alignment_14.sh file \n",
    "   2. Loop through each sample's fastq files to perform alignment and bam conversion/indexing\n",
    "      1. bwa mem is used for mapping medium length PE reads. \n",
    "      2. bam binamry file, sorting and indexing for high performance required by downstream steps. \n",
    "   3. Checking mapped reads \n",
    "      1. samples=(\"C_MB014-2\" \"C_MB014-3\" \"FFPE_C_MB014\" \"PBMC_C_MB014\" \"PBMC_C_MB014-2\")\n",
    "      2. for s in \"${samples[@]}\"; do   bam=\"/Volumes/T7/GTMBioinfo/aligned/${s}.chr12.sorted.bam\";   echo \" Checking $s\";   samtools idxstats \"$bam\";   echo \"\"; done\n",
    "   \n",
    "            Checking C_MB014-2\n",
    "            12      133275309       215463  25540 (chromosome 12, chromosome length, mapped reads, unmapped reads)\n",
    "            *       0       0       513128\n",
    "\n",
    "            Checking C_MB014-3\n",
    "            12      133275309       316406  34464\n",
    "            *       0       0       745550\n",
    "\n",
    "            Checking FFPE_C_MB014\n",
    "            12      133275309       308192  30161\n",
    "            *       0       0       659328\n",
    "\n",
    "            Checking PBMC_C_MB014\n",
    "            12      133275309       331136  39013\n",
    "            *       0       0       811528\n",
    "\n",
    "            Checking PBMC_C_MB014-2\n",
    "            12      133275309       401130  46824\n",
    "            *       0       0       971682\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0fb07cc",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "Variant calling\n",
    "\n",
    "script variant_calling.py\n",
    "script variant_calling_lowerthreshold.py  -- with relaxing thresholds to check if variants can be found. \n",
    "\n",
    "---check variant results\n",
    "\n",
    "cd /Volumes/T7/GTMBioinfo/variants\n",
    "\n",
    "KRAS_REGION=\"12:25245350-25250930\"\n",
    "\n",
    "for s in C_MB014-2 C_MB014-3 FFPE_C_MB014 PBMC_C_MB014 PBMC_C_MB014-2; do\n",
    "  echo \"$s — Filtered VCF:\"\n",
    "  bcftools view -r $KRAS_REGION /Volumes/T7/GTMBioinfo/variants/${s}.filtered.vcf.gz | grep -v \"^#\" || echo \"No variants found in filtered VCF\"\n",
    "  echo \"\"\n",
    "\n",
    "  echo \"$s — Unfiltered VCF:\"\n",
    "  bcftools view -r $KRAS_REGION /Volumes/T7/GTMBioinfo/variants/${s}.unfiltered.vcf.gz | grep -v \"^#\" || echo \"No variants found in unfiltered VCF\"\n",
    "  echo \"------------------------------------------------------------\"\n",
    "done\n",
    "\n",
    "for s in C_MB014-2 C_MB014-3 FFPE_C_MB014 PBMC_C_MB014 PBMC_C_MB014-2; do\n",
    ">   echo \"$s — Filtered VCF:\"\n",
    ">   bcftools view -r $KRAS_REGION /Volumes/T7/GTMBioinfo/variants/${s}.filtered.vcf.gz | grep -v \"^#\" || echo \"No variants found in filtered VCF\"\n",
    ">   echo \"\"\n",
    "> \n",
    ">   echo \"$s — Unfiltered VCF:\"\n",
    ">   bcftools view -r $KRAS_REGION /Volumes/T7/GTMBioinfo/variants/${s}.unfiltered.vcf.gz | grep -v \"^#\" || echo \"No variants found in unfiltered VCF\"\n",
    ">   echo \"------------------------------------------------------------\"\n",
    "> done\n",
    "C_MB014-2 — Filtered VCF:\n",
    "No variants found in filtered VCF\n",
    "\n",
    "C_MB014-2 — Unfiltered VCF:\n",
    "No variants found in unfiltered VCF\n",
    "------------------------------------------------------------\n",
    "C_MB014-3 — Filtered VCF:\n",
    "No variants found in filtered VCF\n",
    "\n",
    "C_MB014-3 — Unfiltered VCF:\n",
    "No variants found in unfiltered VCF\n",
    "------------------------------------------------------------\n",
    "FFPE_C_MB014 — Filtered VCF:\n",
    "No variants found in filtered VCF\n",
    "\n",
    "FFPE_C_MB014 — Unfiltered VCF:\n",
    "No variants found in unfiltered VCF\n",
    "------------------------------------------------------------\n",
    "PBMC_C_MB014 — Filtered VCF:\n",
    "12      25250133        .       G       A       .       clustered_events;haplotype      AS_FilterStatus=SITE;AS_SB_TABLE=0,0|0,0;DP=1;ECNT=7;ECNTH=7;GERMQ=40;MBQ=0,37;MFRL=0,0;MMQ=60,48;MPOS=53;POPAF=7.3;TLOD=4.2        GT:AD:AF:DP:F1R2:F2R1:FAD:PGT:PID:PS:SB     0|1:0,1:0.667:1:0,1:0,0:0,1:0|1:25250133_G_A:25250133:0,0,1,0\n",
    "12      25250135        .       GG      AA      .       clustered_events;haplotype      AS_FilterStatus=SITE;AS_SB_TABLE=0,0|0,0;DP=1;ECNT=7;ECNTH=7;GERMQ=40;MBQ=0,37;MFRL=0,0;MMQ=60,48;MPOS=51;POPAF=7.3;TLOD=4.2        GT:AD:AF:DP:F1R2:F2R1:FAD:PGT:PID:PS:SB     0|1:0,1:0.667:1:0,1:0,0:0,1:0|1:25250133_G_A:25250133:0,0,1,0\n",
    "12      25250144        .       G       A       .       clustered_events;haplotype      AS_FilterStatus=SITE;AS_SB_TABLE=0,0|0,0;DP=1;ECNT=7;ECNTH=7;GERMQ=40;MBQ=0,25;MFRL=0,0;MMQ=60,48;MPOS=42;POPAF=7.3;TLOD=4.2        GT:AD:AF:DP:F1R2:F2R1:FAD:PGT:PID:PS:SB     0|1:0,1:0.667:1:0,1:0,0:0,1:0|1:25250133_G_A:25250133:0,0,1,0\n",
    "12      25250156        .       G       A       .       clustered_events;haplotype      AS_FilterStatus=SITE;AS_SB_TABLE=0,0|0,0;DP=1;ECNT=7;ECNTH=7;GERMQ=40;MBQ=0,37;MFRL=0,0;MMQ=60,48;MPOS=30;POPAF=7.3;TLOD=4.2        GT:AD:AF:DP:F1R2:F2R1:FAD:PGT:PID:PS:SB     0|1:0,1:0.667:1:0,1:0,0:0,1:0|1:25250156_G_A:25250156:0,0,1,0\n",
    "12      25250165        .       A       C       .       base_qual;clustered_events;haplotype    AS_FilterStatus=base_qual;AS_SB_TABLE=0,0|0,0;DP=1;ECNT=7;ECNTH=7;GERMQ=40;MBQ=0,11;MFRL=0,0;MMQ=60,48;MPOS=21;POPAF=7.3;TLOD=4.2   GT:AD:AF:DP:F1R2:F2R1:FAD:PGT:PID:PS:SB     0|1:0,1:0.667:1:0,0:0,0:0,1:0|1:25250156_G_A:25250156:0,0,1,0\n",
    "12      25250169        .       G       A       .       clustered_events;haplotype      AS_FilterStatus=SITE;AS_SB_TABLE=0,0|0,0;DP=1;ECNT=7;ECNTH=7;GERMQ=40;MBQ=0,37;MFRL=0,0;MMQ=60,48;MPOS=17;POPAF=7.3;TLOD=4.2        GT:AD:AF:DP:F1R2:F2R1:FAD:PGT:PID:PS:SB     0|1:0,1:0.667:1:0,1:0,0:0,1:0|1:25250156_G_A:25250156:0,0,1,0\n",
    "12      25250171        .       G       A       .       clustered_events;haplotype      AS_FilterStatus=SITE;AS_SB_TABLE=0,0|0,0;DP=1;ECNT=7;ECNTH=7;GERMQ=40;MBQ=0,37;MFRL=0,0;MMQ=60,48;MPOS=15;POPAF=7.3;TLOD=4.2        GT:AD:AF:DP:F1R2:F2R1:FAD:PGT:PID:PS:SB     0|1:0,1:0.667:1:0,1:0,0:0,1:0|1:25250156_G_A:25250156:0,0,1,0\n",
    "\n",
    "PBMC_C_MB014 — Unfiltered VCF:\n",
    "12      25250133        .       G       A       .       .       AS_SB_TABLE=0,0|0,0;DP=1;ECNT=7;ECNTH=7;MBQ=0,37;MFRL=0,0;MMQ=60,48;MPOS=53;POPAF=7.3;TLOD=4.2      GT:AD:AF:DP:F1R2:F2R1:FAD:PGT:PID:PS:SB 0|1:0,1:0.667:1:0,1:0,0:0,1:0|1:25250133_G_A:25250133:0,0,1,0\n",
    "12      25250135        .       GG      AA      .       .       AS_SB_TABLE=0,0|0,0;DP=1;ECNT=7;ECNTH=7;MBQ=0,37;MFRL=0,0;MMQ=60,48;MPOS=51;POPAF=7.3;TLOD=4.2      GT:AD:AF:DP:F1R2:F2R1:FAD:PGT:PID:PS:SB 0|1:0,1:0.667:1:0,1:0,0:0,1:0|1:25250133_G_A:25250133:0,0,1,0\n",
    "12      25250144        .       G       A       .       .       AS_SB_TABLE=0,0|0,0;DP=1;ECNT=7;ECNTH=7;MBQ=0,25;MFRL=0,0;MMQ=60,48;MPOS=42;POPAF=7.3;TLOD=4.2      GT:AD:AF:DP:F1R2:F2R1:FAD:PGT:PID:PS:SB 0|1:0,1:0.667:1:0,1:0,0:0,1:0|1:25250133_G_A:25250133:0,0,1,0\n",
    "12      25250156        .       G       A       .       .       AS_SB_TABLE=0,0|0,0;DP=1;ECNT=7;ECNTH=7;MBQ=0,37;MFRL=0,0;MMQ=60,48;MPOS=30;POPAF=7.3;TLOD=4.2      GT:AD:AF:DP:F1R2:F2R1:FAD:PGT:PID:PS:SB 0|1:0,1:0.667:1:0,1:0,0:0,1:0|1:25250156_G_A:25250156:0,0,1,0\n",
    "12      25250165        .       A       C       .       .       AS_SB_TABLE=0,0|0,0;DP=1;ECNT=7;ECNTH=7;MBQ=0,11;MFRL=0,0;MMQ=60,48;MPOS=21;POPAF=7.3;TLOD=4.2      GT:AD:AF:DP:F1R2:F2R1:FAD:PGT:PID:PS:SB 0|1:0,1:0.667:1:0,0:0,0:0,1:0|1:25250156_G_A:25250156:0,0,1,0\n",
    "12      25250169        .       G       A       .       .       AS_SB_TABLE=0,0|0,0;DP=1;ECNT=7;ECNTH=7;MBQ=0,37;MFRL=0,0;MMQ=60,48;MPOS=17;POPAF=7.3;TLOD=4.2      GT:AD:AF:DP:F1R2:F2R1:FAD:PGT:PID:PS:SB 0|1:0,1:0.667:1:0,1:0,0:0,1:0|1:25250156_G_A:25250156:0,0,1,0\n",
    "12      25250171        .       G       A       .       .       AS_SB_TABLE=0,0|0,0;DP=1;ECNT=7;ECNTH=7;MBQ=0,37;MFRL=0,0;MMQ=60,48;MPOS=15;POPAF=7.3;TLOD=4.2      GT:AD:AF:DP:F1R2:F2R1:FAD:PGT:PID:PS:SB 0|1:0,1:0.667:1:0,1:0,0:0,1:0|1:25250156_G_A:25250156:0,0,1,0\n",
    "------------------------------------------------------------\n",
    "PBMC_C_MB014-2 — Filtered VCF:\n",
    "No variants found in filtered VCF\n",
    "\n",
    "PBMC_C_MB014-2 — Unfiltered VCF:\n",
    "No variants found in unfiltered VCF\n",
    "------------------------------------------------------------\n",
    "\n",
    "\n",
    "Interpret mapping in PBMC_C_MB014 - DP=1 : total depth - only 1 read supports this site mutation, which cannot support variant statistically significant. \n",
    "\n",
    "IGV screenshot - https://github.com/justin-mbca/GTMBioinfo/blob/main/docs/igv_snapshot.png to inspect sample PBMC_C_MB014 mapping and vcf. \n",
    "\n",
    "  - all these sample has not variants found. \n",
    "\n",
    "\n",
    "Conclusion and Interpretation: \n",
    "\n",
    "  - No true variants were called in filtered or relaxed VCFs\n",
    "\n",
    "  - Unfiltered artifacts in PBMC_C_MB014 were:\n",
    "\n",
    "    - Single-read\n",
    "\n",
    "    - Low base quality\n",
    "\n",
    "    - Not supported in IGV\n",
    "\n",
    "    - no evidence of clonal or subclonal KRAS mutation\n",
    "\n",
    "    - This conclusion is not aligning against the result in the paper \n",
    "\n",
    "        CTDC014 in Table 4 in the paper - https://www.nature.com/articles/s41598-021-95345-4/tables/4\n",
    "        has Gene KRAS missense variant with Pathogenic/Likely_pathogenic "
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
